datafusion/substrait/src/serializer.rs

```diff
@@ -26,28 +26,50 @@ use substrait::proto::Plan;
 
 use std::fs::OpenOptions;
 use std::io::{Read, Write};
+use std::path::Path;
 
-#[allow(clippy::suspicious_open_options)]
-pub async fn serialize(sql: &str, ctx: &SessionContext, path: &str) -> Result<()> {
+/// Plans a sql and serializes the generated logical plan to bytes.
+/// The bytes are then written into a file at `path`.
+///
+/// Returns an error if the file already exists and is not empty.
+pub async fn serialize(
+    sql: &str,
+    ctx: &SessionContext,
+    path: impl AsRef<Path>,
+) -> Result<()> {
     let protobuf_out = serialize_bytes(sql, ctx).await;
-    let mut file = OpenOptions::new().create(true).write(true).open(path)?;
+
+    if std::fs::metadata(path.as_ref()).is_ok_and(|meta| meta.len() > 0) {
+        return Err(DataFusionError::Substrait(format!(
+            "Failed to encode plan: the file {} already exists and is not empty",
+            path.as_ref().display()
+        )));
+    }
+
+    let mut file = OpenOptions::new()
+        .create(true)
+        .write(true)
+        .truncate(true)
+        .open(path)?;
```

Comment:
Looking at this again; we can further simply this with [OpenOptions::create_new](https://doc.rust-lang.org/stable/std/fs/struct.OpenOptions.html#method.create_new).

```suggestion
    let mut file = OpenOptions::new()
        .write(true)
        .create_new(true)
        .open(path)?;
```

This means we also get an error if the file is empty, however this check is atomic, and the existing code is not (the empty check and the open).

Reply:
@mbrobbel  Didn't consider the atomicity, good catch!